pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +278,180 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
+	drivers []storage.CSIDriverInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	driverNamesInSpecs := make(sets.String)
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecNodeID(driver.NodeID, idxPath.Child("nodeID"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoSpecTopologyKeys(driver.TopologyKeys, idxPath.Child("topologyKeys"))...)
+
+		// check for duplicate entries for the same driver in specs
+		if driverNamesInSpecs.Has(driver.Name) {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), driver.Name, fmt.Sprintf("duplicate entries found for driver: %s in driver specs", driver.Name)))
+		}
+		driverNamesInSpecs.Insert(driver.Name)
+		topoKeys := make(sets.String)
+		for _, key := range driver.TopologyKeys {
+			if topoKeys.Has(key) {
+				allErrs = append(allErrs, field.Invalid(idxPath.Child("topologyKeys"), driver.TopologyKeys, fmt.Sprintf("duplicate topology keys %s found for driver %s in driver specs", key, driver.Name)))
+			}
+			topoKeys.Insert(key)
+		}
+
+	}
+
+	return allErrs
+}
+
+// ValidateCSIDriverInfoStatuses tests that the specified CSIDriverInfoStatuses
+// have valid data.
+func validateCSIDriverInfoStatuses(
+	drivers []storage.CSIDriverInfoStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
+		return allErrs
+	}
+
+	driverNamesInStatuses := make(sets.String)
+	for i, driver := range drivers {
+		idxPath := fldPath.Index(i)
+		allErrs = append(allErrs, apivalidation.ValidateCSIDriverName(driver.Name, idxPath.Child("name"))...)
+		allErrs = append(allErrs, validateCSIDriverInfoStatusVolumePluginMechanism(string(driver.VolumePluginMechanism), idxPath.Child("volumePluginMechanism"))...)
+
+		// check for duplicate entries for the same driver in statuses
+		if driverNamesInStatuses.Has(driver.Name) {
+			idxPath := fldPath.Index(i)
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), driver.Name, fmt.Sprintf("duplicate entries found for driver: %s in driver statuses", driver.Name)))
+		}
+		driverNamesInStatuses.Insert(driver.Name)
+	}
+
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecNodeID tests if Name in CSIDriverInfoSpec is a valid node id.
+func validateCSIDriverInfoSpecNodeID(nodeID string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(nodeID) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, nodeID))
+	}
+	return allErrs
+}
+
+// validateCSIDriverInfoSpecTopologyKeys tests if TopologyKeys in CSIDriverInfoSpec are
+// valid topology keys.
+func validateCSIDriverInfoSpecTopologyKeys(keys []string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(keys) == 0 {
```

Comment:
What if I legitimately do not have any topology keys?  Why is this required?

Reply:
Right, this shouldn't be required.  I'll remove this check.